### ISSUE TITLE
About the outputs of PINN

### ISSUE THREAD

**Issue by Ddkowner** [JinshuaiBai/LSWR_loss_function_PINN#5]
Hello @JinshuaiBai ,
I am not majoring in mechanics, so I have some difficulties in understanding formulas. Please help me with the following questions:

1. During the definition and training phase, the last four inputs of PINN are upper, lower, left, and right boundaries; But when predicting, PINN directly passes in the complete xy five times. Can we predict the entire field by passing in xy five times? Why is the code not defined in this way?
2. I noticed that in the outputs of PINN, s_u_x is equal to s_b_x, s_u_y is equal to s_b_y, s_l_x is equal to s_r_x, and s_l_y is equal to s_r_y. Is this related to passing in xy five times? I think both s_u_x and s_l_x represent the x direction, but why are they not equal?
3. Why use s_r_x to represent the stress in the x direction, s_u_y to represent the stress in the y direction, and s_u_x to represent the shear force?

**Comment by JinshuaiBai** [comment]
For your questions:
1. The reason why there are five xy in a list is that the input of a PINN (the class defined in the PINN.py file) needs 5 inputs. In fact, referring to the code in PINN.py file, if you only want to output stress fields, just make sure the first element in your input list is xy, while the other four inputs can be random arrays as long as they satisfy the shape requirement (,2).
2. No, s_u_x is equal to s_u_x and s_u_y is equal to s_b_y. Please check lines 72-80 in the OPT.py file.
3. For the right boundary, the traction boundary condition in the x direction ${t_x}$ is equal to the normal stress in the x direction ${\sigma_x}$, while the traction boundary condition ${t_y}$ in the y direction is equal to the shear stress ${\tau_{xy}}$. Here, for simplicity, I directly use x and y to represent the boundary condition's directions rather than the components of stresses.

You may need to further strengthen your coding skills.
Hope this can answer your questions!

**Comment by Ddkowner** [comment]
Thank you very much for your reply! In order not to disturb you, I came to ask the question after studying the code carefully. My description was not very clear, let me put it another way.

1. Do you intend to extend the boundary to the entire object in order to obtain the full stress/strain fields by passing xy five times into pinn. predict()? Or is it to follow the rule that model.predict() must pass in inputs with the same batchsize? 

2. I understand that in OPT.py, it is calculating the error between the actual value and the predicted value of pin. I found that Sux=Sbx, Suy=Sby, Slx=Srx, Sly=Sry when comparing the pinn predicted results. Is this because we set all boundaries to xy, so there is no difference between u/b and l/r?
 
3. Thank you very much for your response to question 3, which helped me understand how to select variables to represent stress. In addition to the equality relationship mentioned above, I found that as shear stress, Sux and Sly are equal, but Slx and Sby are not equal. Can you further explain the relationship between these two?

![image](https://github.com/JinshuaiBai/LSWR_loss_function_PINN/assets/169640000/79e92f27-d300-4111-819c-526abd0e08b8)

Thanks a lot for your help!

**Comment by JinshuaiBai** [comment]
> Thank you very much for your reply! In order not to disturb you, I came to ask the question after studying the code carefully. My description was not very clear, let me put it another way.
> 
> 1. Do you intend to extend the boundary to the entire object in order to obtain the full stress/strain fields by passing xy five times into pinn. predict()? Or is it to follow the rule that model.predict() must pass in inputs with the same batchsize?
> 2. I understand that in OPT.py, it is calculating the error between the actual value and the predicted value of pin. I found that Sux=Sbx, Suy=Sby, Slx=Srx, Sly=Sry when comparing the pinn predicted results. Is this because we set all boundaries to xy, so there is no difference between u/b and l/r?
> 3. Thank you very much for your response to question 3, which helped me understand how to select variables to represent stress. In addition to the equality relationship mentioned above, I found that as shear stress, Sux and Sly are equal, but Slx and Sby are not equal. Can you further explain the relationship between these two?
> 
> ![image](https://private-user-images.githubusercontent.com/169640000/342617227-79e92f27-d300-4111-819c-526abd0e08b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzNjg2MjIsIm5iZiI6MTcxOTM2ODMyMiwicGF0aCI6Ii8xNjk2NDAwMDAvMzQyNjE3MjI3LTc5ZTkyZjI3LWQzMDAtNDExMS04MTljLTUyNmFiZDBlMDhiOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyNlQwMjE4NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZTgyYmQ3NDliZTg5MzExNDJiYTg2MWQzY2NmNzVlMmI0NDlhNjRlYWUwNjVmMWY4MTY2YTBiYzRmYjcxODQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.5tX9tTsQOqeXUiiDZeu1ClENf_bfakAKziDvuQyqBd8)
> 
> Thanks a lot for your help!

For your questions:
1. The latter one is correct. To strictly follow the input rule in the model.predict(), we compressed five xy in a list as the input. The number of elements in the input should align with the number of input elements defined in "PINN.py files",
2. From your description, I guess you are running the pure bending case, aren't you? If so, only for this case, the traction boundary conditions are symmetrical. If you try the other cases, for example, the perforated plate problem, the stress fields on the boundaries are no longer symmetry. In summary, the stress solution depends on problems in terms of geometry, boundary conditions, and material properties.
3. Please refer to my previous response.

I will also recommend you to read some textbooks regarding basic mechanics books. They will help you better understand the formula used in the loss function and the results.

Hope these can answer your questions.

**Comment by zhangan810** [comment]
Hello @JinshuaiBai ,

Thank you for sharing the code. I also have this issue, but I believe the reason why the PINN directly passes in the complete xy five times is that it calculate s_u_x,s_u_y  etc., are not only the stress values at the boundary but at all xy coordinates.  This is because, when computing partial derivatives, the points used are not just those on the boundary but all points., since xy is used. This is also why Sux=Sbx, Suy=Sby, Slx=Srx, Sly=Sry. I'm not sure if my understanding is correct.

**Comment by JinshuaiBai** [comment]
> Hello @JinshuaiBai ,
> 
> Thank you for sharing the code. I also have this issue, but I believe the reason why the PINN directly passes in the complete xy five times is that it calculate s_u_x,s_u_y etc., are not only the stress values at the boundary but at all xy coordinates. This is because, when computing partial derivatives, the points used are not just those on the boundary but all points., since xy is used. This is also why Sux=Sbx, Suy=Sby, Slx=Srx, Sly=Sry. I'm not sure if my understanding is correct.

No. The reason why five xy are fed into the PINN is to follow the input size in the model.predict(), which is defined in the PINN.py file (as shown below). 
![image](https://github.com/user-attachments/assets/2f9daadd-181e-4951-bdfa-5fee02127514)
In the training process, those inputs are all used to calculate the residuals from the equilibrium equation and the boundary conditions. 
In the prediction stage, we need to feed the input of our PINN with the same size. In fact, only the first xy is required, while the following four xy are not accurately outputted. You can do this, create a list with five elements and make sure the first one is xy. As for the rest elements, you can use randomly generated four arrays and make sure their size is (n,2). Note that n can be any positive integer. You will still get the same results you need.

Hope this can answer your question.
